api/paidAction/index.js

```diff
@@ -30,75 +34,166 @@ export const paidActions = {
   TERRITORY_BILLING,
   TERRITORY_UNARCHIVE,
   DONATE,
+  TRANSFER,
   BUY_CREDITS
 }
 
+async function checkUser (userId, { models, tx, me }, minSats, minCreds) {
+  // count pending invoices and bail if we're over the limit
+  const pendingInvoices = await ((tx ?? models).invoice.count({
+    where: {
+      userId,
+      actionState: {
+        // not in a terminal state. Note: null isn't counted by prisma
+        notIn: PAID_ACTION_TERMINAL_STATES
+      }
+    }
+  }))
+
+  console.log('pending paid actions', pendingInvoices)
+  if (pendingInvoices >= MAX_PENDING_PAID_ACTIONS_PER_USER) {
+    throw new Error('you have too many pending paid actions, cancel some or wait for them to expire')
+  }
+
+  if (minSats != null && toPositiveBigInt(minSats) > (me?.msats ?? 0n)) {
+    throw new Error('you do not have enough sats to perform this action')
+  }
+  if (minCreds != null && toPositiveBigInt(minCreds) > (me?.mcredits ?? 0n)) {
+    throw new Error('you do not have enough credits to perform this action')
+  }
+}
+
 export default async function performPaidAction (actionType, args, context) {
   try {
-    const { me, models, forceFeeCredits } = context
-    const paidAction = paidActions[actionType]
-
+    context = { ...context }
     console.group('performPaidAction', actionType, args)
-
+    const paidAction = paidActions[actionType]
     if (!paidAction) {
-      throw new Error(`Invalid action type ${actionType}`)
+      throw new Error(`ynvalid action type ${actionType}`)
     }
 
-    if (!me && !paidAction.anonable) {
-      throw new Error('You must be logged in to perform this action')
-    }
+    context.me = context.me ? await context.models.user.findUnique({ where: { id: context.me.id } }) : undefined
+    context.actionAttempt = context.actionAttempt ?? 0
+    context.forceInternal = context.forceInternal ?? false
```

Comment:
this is currently used by things like auto territory renewal to force the code to use only internal payments (fee credits and reward sats)